cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractGoal.java

```diff
@@ -238,12 +252,14 @@ protected Broker maybeApplyBalancingAction(ClusterModel clusterModel,
    * @param candidateReplicasToSwapWith Candidate replicas from the same destination broker to swap in the order of
    *                                    attempts to swap.
    * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization -- e.g. excluded brokers for replica move.
    * @return True the swapped in replica if succeeded, null otherwise.
    */
   Replica maybeApplySwapAction(ClusterModel clusterModel,
                                Replica sourceReplica,
                                SortedSet<Replica> candidateReplicasToSwapWith,
-                               Set<Goal> optimizedGoals) {
+                               Set<Goal> optimizedGoals,
+                               OptimizationOptions optimizationOptions) {
```

Comment:
Agreed that the relevant logic can be moved inside `getEligibleReplicasForSwap`. Regarding the logic, it should be sufficient to check the swap one way, because candidate replicas are chosen to be among the ones that are eligible to be received by this broker.